### ISSUE TITLE
Progress update: 2026-03-11 16:22

### ISSUE THREAD

**Issue by pierretokns** [pierretokns/agenthub#25]
### Summary

9 experiments completed on the MLX-native pipeline. First successful full training run (exp-11) achieved **new best primary score: 32.67**, up from 27.20 baseline. Exp-12 currently training, targeting clustering improvement.

### Experiment Counts

| Status | Count |
|---|---|
| Keep | 2 |
| Discard | 1 |
| Crash | 3 |
| Contaminated | 3 |
| **Total** | **9** |

### Current Best: exp-11 (primary = 32.67)

**Config:** proj_dim=768, hard-neg mining (8K docs, batch_size=32), 4-stage training (warmup 15min + contrastive 30min + mining + finetune 30min), reddit + SE + NQ + MSMARCO data, 134K clean pairs.

| Task | Category | Score | vs Baseline (exp-3) |
|---|---|---|---|
| STSBenchmark | STS | 54.73 | +9.0 |
| SICK-R | STS | 51.77 | +2.5 |
| TwitterURLCorpus | PairClass | 69.27 | +1.2 |
| SprintDuplicateQuestions | PairClass | 60.91 | +26.8 |
| TwentyNewsgroupsClustering | Clustering | 11.25 | +2.1 |
| RedditClustering | Clustering | 6.85 | -2.0 |
| SciFact | Retrieval | 9.37 | +5.3 |
| NFCorpus | Retrieval | 3.07 | +0.8 |

**Primary score breakdown:**
- STS avg: 53.25 (weight 0.3)
- PairClassification avg: 65.09 (weight 0.2)
- Clustering avg: 9.05 (weight 0.2)
- Retrieval avg: 6.22 (weight 0.3)

**Biggest gaps to close:** Clustering (~9%) and Retrieval (~6%) are dragging the score down. Target competitor gte-modernbert-base is ~65 overall.

### Current Activity

**Exp-12 in progress:** `add ag_news same-label pairs for clustering + finetune temp 0.02→0.05; target clustering(9%→15%)`
- Contrastive stage ~83% complete (step 1250, loss 0.250, 1501s/1800s)
- Contrastive loss already below exp-11's final (0.250 vs 0.267)
- ETA for scores: ~20 min

### Key Milestones This Session

1. **MLX migration complete** — training pipeline fully native Apple Silicon, no PyTorch
2. **First successful full run** (exp-11) — survived all 4 stages including hard-neg mining
3. **OOM resolved** — mining limited to 8K docs + batch_size=32
4. **Stage checkpointing added** — `--resume-stage` to skip completed stages on crash
5. **program.md rewritten** with SOTA research directions (latent attention pooling, Matryoshka, zELO distillation, multi-layer pooling)
6. **Data pipeline optimized** — streaming for large datasets, local JSON caching, hf_transfer

### Errors & Issues

- **exp-8/9/10 OOM crashes** during hard-neg mining stage — resolved by limiting to 8K docs
- **Log buffering** — fixed with readline() loop and flush=True
- `mx.metal.clear_cache` deprecation warning (cosmetic, no impact)
- Agent accumulates zombie `sleep` processes for monitoring (harmless, cleaned on session restart)

### System Health

- **Memory:** 64GB unified, ~900MB RSS for train.py during training, ~4.5GB during MTEB eval
- **Supervisor:** `com.gourmand.embed-research` running (PID 85973, exit 0)
- **Agenthub server:** `com.gourmand.agenthub-server` running (PID 55289, exit 0)
- **Claude agent:** Sonnet 4.6, --max-turns 200, active since 12:54 PM
- **wandb:** Logging to gourmand-labs/autoresearch-embed

### Agenthub Activity (last 5 messages)

```
[1] Agent online (03:21)
[2] CRASH test wrapper (14:03)
[3] DISCARD 32.69 test wrapper MPS (15:04)
[4] CRASH exp-8 OOM (17:04)
[5] CRASH exp-9 OOM (17:06)
[6] CRASH exp-10 OOM (17:57)
[7] KEEP 32.67 exp-11 (19:29)
```

### Next Steps (autonomous)

The agent will continue the research loop after exp-12. Key research directions from updated program.md:
- Latent attention pooling (NV-Embed-v2 technique)
- Multi-layer trainable pooling
- Matryoshka Representation Learning
- Better data stratification for clustering/retrieval weakness
- Temperature and loss function sweeps
